File: code/12-JavaCodeLoop/JavaTDDLoop/src/test/java/com/example/testsupport/RecordingNumberGuesser.java
package com.example.testsupport;

import com.example.valueobjects.Interval;
import com.example.domain.NumberGuesser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Test spy for {@code NumberGuesser}.
 *
 * <p>Records every interval it is asked to guess against, then delegates to the real
 * {@code NumberGuesser} strategy so behaviour stays realistic. Used to observe how the
 * interval narrows across a {@code GuessGame} play-through.</p>
 */
public class RecordingNumberGuesser extends NumberGuesser {

    private final List<Interval> recordedIntervals = new ArrayList<>();

    @Override
    public int nextGuess(Interval interval) {
        recordedIntervals.add(interval);
        return super.nextGuess(interval);
    }

    public List<Interval> recordedIntervals() {
        return Collections.unmodifiableList(recordedIntervals);
    }
}

File: code/13-Java2Python/python/tests/testsupport/recording_number_guesser.py
from domain.number_guesser import NumberGuesser
from valueobjects.interval import Interval


class RecordingNumberGuesser(NumberGuesser):
    """Test spy for NumberGuesser.

    Records every interval it is asked to guess against, then delegates to
    the real NumberGuesser strategy so behaviour stays realistic. Used to
    observe how the interval narrows across a GuessGame play-through.
    """

    def __init__(self) -> None:
        self._recorded_intervals: list[Interval] = []

    def next_guess(self, interval: Interval) -> int:
        self._recorded_intervals.append(interval)
        return super().next_guess(interval)

    @property
    def recorded_intervals(self) -> list[Interval]:
        return list(self._recorded_intervals)
